### DISCUSSION TITLE
关于#320的卡死问题

### DISCUSSION THREAD

**Discussion by CinnabarCHU** [mikumifa/biliTickerBuy#359] [聊天专区]
不知为何，可能是我的使用习惯问题，我总是会遇到程序卡死问题，#301 的卡死基本已经确定原因并且给出修改方案了，但是#320 这里提到的卡死，虽然我现在怀疑是题主开了手动抢票而不自知，但是确实也可能是gradio前端网页没有按照预期正常更新导致的彻底卡死，这种方法目前来看可以通过点击“停止”按钮（如果存在的话）来解除。问题恰恰在于，如果前端网页没有按照预期正常更新，“停止”按钮将不会显示，而由于程序处于挂起状态，即使刷新网页，停止按钮的`visible`属性仍然无法更新为`True`，这将导致程序如同之前无法取消的计时一样彻底卡死。

此外，验证码模式是随着单选框的更新而更新的，点击开始抢票按钮并不会检查前端选择的验证码模式，一旦后端崩溃重启`select_way`重置为`0`，由于前端不会随之自动刷新，那么此时若用户在重新连接后直接点击开始抢票，将导致用户虽然选择了其他验证码模式，但实际仍然启用手动过码模式。我本人对gradio不熟悉，不过目前观察到的情况是，在后端重启而前端不刷新的情况下，前端可以对后端发出指令，但是后端对前端网页的更新却无法回传，这意味着用户随时可能面临前端挂起卡死不得不重启程序的风险。

我对于gradio的了解有限，不知道在后端重启后前端仅能保持单向通信是否是gradio的特性，也不清楚在浏览器刷新后重新更新网页前端的开销，我个人认为避免前端引起卡死导致重启整个程序的方法是添加一个时刻显示的停止按钮，并保证点击停止按钮后前端网页的Go Tab可以得到一次完全的更新。不知道贡献者们是否觉得这个问题是否需要解决？若需要解决的话是否有更好的方案？

**Comment by mikumifa** [comment]
> 前端不会随之自动刷新，那么此时若用户在重新连接后直接点击开始抢票

这个问题不好解决，因为程序是有状态的，状态都在后端，后端崩溃之后，前端应该重启

> 前端网页没有按照预期正常更新，“停止”按钮将不会显示

v2.8.5-patch.1 最新版本加了停止按钮,可以测试一下

**Comment by gpc123456** [comment]
`v2.8.5-patch.1` 更新后只要后端没有卡死，前端应该是可以正常停止并显示相应的状态了，不会像以前一样抢票过程中点击停止但没有作用了。但如果不幸后端卡死了(目前还没发现这种情况)，前端点击停止向后端发送请求也是没有用的，这时候还是手动终止终端后重启脚本吧。

由于在Gradio中刷新页面需要使用 `yield` 来切换线程，这是一笔很大的性能开销，所以在抢票过程中并没有进行前端页面的刷新，应该按照输出框内的提示，通过观察控制台的log来判断当前的抢票状况，必要时手动重启脚本。当前在抢票过程中，通过检测 `isRunning` 这个变量来确定是否应该停止抢票，当停止按钮被触发时会将其设置为0，进而停止抢票过程。

Gradio进程重启后原有的浏览器会话将不再可靠，应该手动刷新重载浏览器页面建立新会话，Gradio目前好像暂时无法控制浏览器在原有页面建立新的连接，或许可以像`stable-diffusion-webui`那样，当控制台重启时，自动弹出一个新的浏览器窗口给用户。

Edit：
如这个 [Fork](https://github.com/gpc123456/biliTickerBuy)，在Gradio启动时自动打开一个新浏览器页面，这样即使后端重启了，也可以保证用户使用的是最新的会话，避免前端后端因会话不一致导致的数据同步问题。

Update：
已将新建了一个PR #367 ， 应该能够解决这个问题。不过如果使用Linux或MacOS在doker中运行时，是否能够成功直接调用浏览器打开页面还需要测试

Update：
我在阿里云的Ubuntu Linux服务器上用docker跑了一下，貌似在docker中无法调用本地浏览器打开页面，所以如果在Linux或MacOS上运行，还是要关注控制台的状况，如后端发生重启一定要手动刷新一下前端页面。
